### PR TITLE
uncap additional expenses and allow delete

### LINKED ISSUE #5773: Uncap submissions for Additional Expenses and improve UX
## What type(s) of user does this feature affect?
 - supervisors
 - admins
 - volunteers

## Description

If an org has enabled additional expenses the last step of the case contact form (`case_contacts/:id/form/expenses`) allows the volunteer to add additional expenses.

Image

Currently there is a cap of 10 expenses. When you hit 10 the button remains but you cannot add any more.

Image

We have https://www.stimulus-components.com/docs/stimulus-rails-nested-form/ in the application. It could potentially be of use (unless needed I would suggest using the non-extended version of the stimulus controller). 

**Completion Criteria:**

- [ ] Use some javascript to uncap the amount of expenses 
- [ ] Add a way to remove expenses.

## How to access the QA site
_Login Details:_  
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin
- all_casa_admin1@example.com view site as an all casa admin
  - go to `/all_casa_admins/sign_in`  

password for all users: 12345678  

### Questions? Join Slack!

We highly recommend that you join us in [slack](https://join.slack.com/t/rubyforgood/shared_invite/zt-21pyz2ab8-H6JgQfGGI0Ab6MfNOZRIQA) #casa channel to ask questions quickly. And [discord](https://discord.gg/qJcw2RZH8Q) for office hours (currently Tuesday 5-7pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by shacon**
### What github issue is this PR for, if any?
Resolves #5773 

### What changed, and _why_?
Allows for adding more than 10 additional expenses and allows for deleting additional expenses from the form
Used this stimulus controller for nested form fields as suggested in the ticket: https://www.stimulus-components.com/docs/stimulus-rails-nested-form/

### How is this **tested**? (please write tests!) 💖💪
Edited the capybara tests that existed with the changes to allow for more than 10 additional expenses. 

Also added capybara test for the delete functionality.

This pr does add a javascript controller that resets the  indexes on the nested fields generated by the stimulus controller. The capybara tests so end up testing this functionality but if it seems like we want specific javascript tests I can take a stab at those.

### Screenshots please :)
Screen Shot 2024-07-23 at 1 15 48 PM

Screen Shot 2024-07-23 at 1 14 26 PM

A note on testing locally - this requires a feature flag to be set and for a config to be set on the related Casa Org. Here's what I ran in the rails console to do this:
```
org =CaseContact.last.casa_case.casa_org

 org.additional_expenses_enabled=true
org.save

Flipper.enable :show_additional_expenses
```

### Feelings gif (optional)
optimistic gif

### CODE REVIEW

File: app/javascript/controllers/additional_expenses_form_controller.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+      container.querySelectorAll('label').forEach(label => {
+        const forAttr = label.getAttribute('for')
+        if (forAttr) {
+          const newForAttr = forAttr.replace(/additional_expenses_attributes_\d+/, `additional_expenses_attributes_${index}`)
+          label.setAttribute('for', newForAttr)
+        }
+      })
+    })
+  }
```
Comment (elasticspoon) [inline app/javascript/controllers/additional_expenses_form_controller.js:35]: What happens if we don't reindex the fields?

Comment (schoork) [reply app/javascript/controllers/additional_expenses_form_controller.js:35]: You shouldn't need a new controller at all for this. This should be doable 100% in the template and controller. Look at the documentation [here](https://www.stimulus-components.com/docs/stimulus-rails-nested-form). 

We already have the component installed, so you can skip the installation step.

File: app/assets/stylesheets/pages/case_contacts.scss [hunk 2]
```diff
@@ -113,6 +113,10 @@ legend {
     cursor: pointer;
 }
 
+.remove-expense-button {
```
Comment (schoork) [inline app/assets/stylesheets/pages/case_contacts.scss:116]: Is this doable through Bootstrap code?

File: app/javascript/controllers/additional_expenses_form_controller.js (hunk 1 above)
Comment (schoork) [reply app/javascript/controllers/additional_expenses_form_controller.js:35]: Essentially, persisted associations get their ID as their index and new records get the timestamp they were created ([line 19](https://github.com/stimulus-components/stimulus-rails-nested-form/blob/master/src/index.ts)). This ensures unique identifiers for every association in the form without needing a reindex.

File: app/javascript/controllers/additional_expenses_form_controller.js
Comment (schoork) [inline app/javascript/controllers/additional_expenses_form_controller.js:1]: You will be able to remove this file.

File: app/javascript/controllers/index.js [hunk 3]
```diff
@@ -14,6 +14,8 @@ import DismissController from './dismiss_controller'
 
 import CourtOrderFormController from './court_order_form_controller'
 
+import AdditionalExpensesFormController from './additional_expenses_form_controller'
```
Comment (schoork) [inline app/javascript/controllers/index.js:17]: This you can remove.

File: app/javascript/controllers/index.js [hunk 4]
```diff
@@ -34,6 +36,7 @@ application.register('autosave', AutosaveController)
 application.register('disable-form', DisableFormController)
 application.register('dismiss', DismissController)
 application.register('court-order-form', CourtOrderFormController)
+application.register('additional-expenses-form', AdditionalExpensesFormController)
```
Comment (schoork) [inline app/javascript/controllers/index.js:39]: Remove as well.

**Review by schoork — CHANGES_REQUESTED** [review]
Overall this looks great, but we don't need the added controller.

File: app/assets/stylesheets/pages/case_contacts.scss (hunk 2 above)
Comment (shacon) [reply app/assets/stylesheets/pages/case_contacts.scss:116]: @schoork I can look into that, haven't worked with bootstrap in a long time but I'll check it out

File: app/javascript/controllers/additional_expenses_form_controller.js (hunk 1 above)
Comment (shacon) [reply app/javascript/controllers/additional_expenses_form_controller.js:35]: @schoork @elasticspoon thanks! I'll make these changes and push something up by the end of this week

File: app/javascript/controllers/index.js [hunk 5]
```diff
@@ -42,3 +45,4 @@ application.register('select-all', SelectAllController)
 application.register('sidebar', SidebarController)
 application.register('sidebar-group', SidebarGroupController)
 application.register('truncated-text', TruncatedTextController)
+application.register('nested-form', NestedForm)
```
Comment (shacon) [inline app/javascript/controllers/index.js:48]: I wasn't able to access the nested form controller until I added this - not sure if there is a more standard way to do this. I realize the rest of the imports in here are all local

Comment (schoork) [reply app/javascript/controllers/index.js:48]: The `court_order_form_controller.js` uses it, too. But it loads it directly. This is the right thing to do for your use case.

File: app/views/shared/_additional_expense_form.html.erb
Comment (schoork) [inline app/views/shared/_additional_expense_form.html.erb:1]: It's not a huge deal, but I'd say get rid of the `a` and `b`. With multiple expenses it becomes confusing. 
Screenshot 2024-07-23 at 6 59 24 AM

File: app/views/shared/_additional_expense_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    <li>
+      <div class="expense-container row" data-index="<%= form.index %>">
+        <div class="input-style-1 col-sm-3">
+          <h5><%= form.label :additional_expense, "a. Expense amount" %></h5>
+          <div class="input-group">
+            <span class="dollar-sign">
+              <%= form.number_field :other_expense_amount,
+                placeholder: "0",
+                class: "form-control other-expense-amount",
+                min: "0", max: 1000, step: 0.01 %>
```
Comment (schoork) [inline app/views/shared/_additional_expense_form.html.erb:11]: After saving and coming back to edit, I get an expense in the form with only one cent place. 
Screenshot 2024-07-23 at 7 01 16 AM

**Review by schoork — CHANGES_REQUESTED** [review]
This looks great @shacon. One tweak to the way the cost is displayed when editing and good to go!

File: app/views/shared/_additional_expense_form.html.erb (hunk 6 above)
Comment (shacon) [reply app/views/shared/_additional_expense_form.html.erb:12]: @schoork added number_with_precision to the field, let me know if this display is ok:
Screen Shot 2024-07-23 at 1 12 11 PM

File: app/views/shared/_additional_expense_form.html.erb
Comment (shacon) [reply app/views/shared/_additional_expense_form.html.erb:1]: ah, just saw this, will take those out

**Comment by shacon** [comment]
> This looks great @shacon. One tweak to the way the cost is displayed when editing and good to go!

thanks @schoork! I made the requested changes and updated the screenshots - let me know if there is anything else I should do on my end.

**Review by schoork — APPROVED** [review]
Looks great @shacon! Thanks for the reminder.

**Pull request merged**